src/snowflake/cli/api/utils/definition_rendering.py

```diff
@@ -0,0 +1,237 @@
+    return _get_referenced_vars(ast, template_str)
+
+
+def _build_dependency_graph(
+    env: Environment, all_vars: set[TemplateVar], context: dict[str, Any]
+) -> Graph[TemplateVar]:
+    dependencies_graph = Graph[TemplateVar]()
+    for variable in all_vars:
+        dependencies_graph.add(Node[TemplateVar](key=variable.get_key(), data=variable))
+
+    for variable in all_vars:
+        if variable.is_env_var and variable.get_env_var_name() in os.environ:
+            # If variable is found in os.environ, then use the value as is
+            # skip rendering by pre-setting the rendered_value attribute
+            env_value = os.environ.get(variable.get_env_var_name())
+            variable.rendered_value = env_value
+            variable.templated_value = env_value
+        else:
+            value: Any = variable.read_from_context(context)
+            variable.templated_value = str(value)
+            dependencies_vars = _get_referenced_vars_from_str(
+                env, variable.templated_value
+            )
+            if len(dependencies_vars) == 0:
+                variable.rendered_value = value
+
+            for referenced_var in dependencies_vars:
+                dependencies_graph.add_directed_edge(
+                    variable.get_key(), referenced_var.get_key()
+                )
+
+    return dependencies_graph
+
+
+def _render_graph_node(jinja_env: Environment, node: Node[TemplateVar]) -> None:
+    if node.data.rendered_value is not None:
+        # Do not re-evaluate resolved nodes like env variable nodes
+        # which might contain template-like values, or non-string nodes
+        return
+
+    current_context: dict[str, Any] = {}
+    for dep_node in node.neighbors:
+        dep_node.data.add_to_context(current_context)
+
+    template = jinja_env.from_string(node.data.templated_value)
+    node.data.rendered_value = template.render(current_context)
+
+
+def _render_dict_element(
+    jinja_env: Environment, context: dict[str, Any], element: str
+) -> str:
+    if _get_referenced_vars_from_str(jinja_env, element):
+        template = jinja_env.from_string(element)
+        return template.render(context)
+    return element
+
+
+def render_definition_template(definition: dict[str, Any]) -> dict[str, Any]:
+    if "definition_version" not in definition or Version(
+        definition["definition_version"]
+    ) < Version("1.1"):
+        return definition
+
+    jinja_env = get_snowflake_cli_jinja_env()
+    pdf_context = {CONTEXT_KEY: definition}
+
+    referenced_vars = set()
+
+    def find_any_template_vars(element):
+        referenced_vars.update(_get_referenced_vars_from_str(jinja_env, element))
+
+    traverse(definition, visit_action=find_any_template_vars)
+
+    dependencies_graph = _build_dependency_graph(
+        jinja_env, referenced_vars, pdf_context
+    )
+
+    def on_cycle_action():
+        raise CycleDetectedError("Cycle detected in templating variables")
```

Comment:
it would be useful for a user to know where the cycle is. If even one of the variables is part of the output, they'd know where to start looking. This is a bit too opaque.